doc/user-guide.md

```diff
@@ -1 +1,380 @@
+When you attempt to merge two commits containing different SHAs for the same
+submodule, Git simply flags the submodule as being conflicted.  This result
+makes sense based on the original use for submodules (on one branch we
+referenced one version of an external repository, on another branch we
+referenced a different version); we need a human to decide which one to use.
+While it might make sense to blindly accept a change to a commit that is a
+descendant of the original, it probably is not reasonable to e.g., create a
+merge commit -- this submodule is an external repository that we may not own.
+
+Unfortunately, this behavior renders Git commands that perform merges:
+`cherry-pick`, `merge`, `rebase`, etc. mostly useless when interacting with a
+mono-repo.  Instead of triggering a conflict when encountering submodule
+differences, we need these commands to:
+
+- `cherry-pick` -- perform a merge in the submodule
+- `merge` -- perform a merge in the submodule
+- `rebase` -- rebase submodule commits
+
+Furthermore, it may be necessary to fetch target commits during these
+operations.
+
+### Workarounds
+
+Workarounds are direct implementation of existing Git commands that provide no
+value other than to work, or to work faster.  One current example is
+`git submodule status`, that outputs the status of only a few
+submodules per second, making it unusable in repositories containing thousands
+of submodules.
+
+## Client-side Model
+
+### Overview
+
+To the greatest extent possible, we want to allow users to treat the
+combination of meta-repo + sub-repos as a mono-repo.  The goal is to make it
+easier to use the mono-repo, not to hide or "protect" the user from the actual
+submodule-based implementation.  We make no attempt to completely wrap the Git
+UI, so users will need to interact directly with submodules from time-to-time,
+e.g., to address merge conflicts.
+
+Conceptually, we view the entire set of (open) sub-repos as a single
+repository, with one:
+
+- current branch -- the current branch in the meta-repo
+- HEAD -- the head of meta-repo
+- working directory -- the tree of all open sub-repos
+- index -- the combined indices of all open sub-repos
+
+Some implications:
+
+- Unless explicitly specified (and then only very rarely), "committishes"
+  (reference names, SHAs, etc.) provided to git-meta commands are resolved only
+  in the meta-repo.
+- The HEADs of submodules are adjusted only as a consequence of changing the
+  HEAD of the meta-repo.
+- Paths in git-meta commands are allowed to refer to any part of the working
+  tree, regardless of current working directory.
+- The working tree and index of the meta-repo itself are ignored unless the
+  user explicitly needs to add files to it. 
+- New commits in submodules (not yet reflected in the HEAD of the meta-repo)
+  are considered to be staged in the logical index of the mono-repo (and show
+  that way in `git meta status`).
+
```

Comment:
Though if anything it probably deserves (more) mention in the architecture.md file.